todo_app/app.py

```diff
@@ -24,37 +33,85 @@ def create_app():
             )
         app.logger.addHandler(handler)
 
+    @login_manager.unauthorized_handler
+    def unauthenticated():
+        uri = oauth_client.prepare_request_uri('https://github.com/login/oauth/authorize')
+        return redirect(uri)
+
+    @login_manager.user_loader
+    def load_user(user_id):
+        return Authenticated_User(user_id)
+
+    @app.route('/login/callback', methods = ['GET'])
+    def login_callback():
+        token_url, headers, body = oauth_client.prepare_token_request(
+            'https://github.com/login/oauth/access_token',
+            authorization_response = request.url,
+            code = request.args.get('code')
+            )
+        
+        token_response = requests.post(
+            token_url, 
+            headers=headers, 
+            data=body, 
+            auth=(oauth_client_id, oauth_client_secret)
+            )
+
+        oauth_client.parse_request_body_response(token_response.text)
+
+        userinfo_endpoint = 'https://api.github.com/user'
+        uri, headers, body = oauth_client.add_token(userinfo_endpoint)
+        user_response = requests.get(uri, headers=headers, data=body)
+        authenticated_user = Authenticated_User(user_response.json()['login'])
+        login_user(authenticated_user)
+
+        return redirect('/')
 
     @app.route('/')
+    @login_required
     def index():
         app.logger.info("Start rendering template for get call")
         app.logger.info(f'Env variables are {os.environ.get("MONGO_USER_NAME")}')
         cards = fetch_all_cards("To Do", "Doing", "Done")  
         view_model_obj = ViewModel(cards) 
+        if app.config.get('LOGIN_DISABLED') or current_user.role == 'writer':
+            user_role = 'writer'
```

Comment:
You don't set `user_role` if this `if` expression isn't true, so it errors on line 85!

Reply:
Ah ! Thanks for the feedback - stupid error on my part